__tests__/Relude_IO_test.re

```diff
@@ -2,6 +2,9 @@ open Jest;
 open Expect;
 open Relude.Globals;
 
+[@coverage exclude_file];
+afterAll(Bisect.Runtime.write_coverage_data);
```

Comment:
Argh :/ I think we need some better way of doing this globally in a Jest project. The problem I ran into is that when Jest exits, it somehow doesn't trigger `process.on('exit')` hooks.

Also, a near-future version of Bisect will support excluding files by pattern in the `ppx-flags`. In the meantime, you could use an exclusions file: https://github.com/aantron/bisect_ppx/blob/master/doc/advanced.md#Excluding

Reply:
> In the meantime, you could use an exclusions file: https://github.com/aantron/bisect_ppx/blob/master/doc/advanced.md#Excluding

I've tried without success, could you please advise?

I added a `dune` file in the root of project:

```
(preprocess (pps bisect_ppx --exclusions bisect.exclude))
(preprocessor_deps (file bisect.exclude))
```

And added a `bisect.exclude` file next to `dune`:

```
file regexp "*_test.bs.js"
file regexp "node_modules/*"
```

Reply:
I understood that you are building with BuckleScript, so this should go into `bsconfig.json`, not a `dune` file. I think (haven't tried) you can pass the option like this:

```
"ppx-flags": [["bisect_ppx/ppx", "--exclusions bisect.exclude"]]
```

[like so](https://github.com/BuckleScript/bucklescript/blob/2e4c0558a8b417983a4c8ce8d6016ea3a4fa1fc5/jscomp/build_tests/scoped_ppx/bsconfig.json).

Then, the file patterns are regexps, not globs, so

```
file regexp ".*_test.bs.js"
```

and you won't be able to exclude `node_modules` from your project, because instrumentation of bs-bastet is controlled from the upstream `node_modules/bs-bastet/bsconfig.json`, and is best disabled by https://github.com/BuckleScript/bucklescript/issues/3761.

Reply:
it seems like bisect has not supported relative path of exclusion file:

```js
"ppx-flags": [["bisect_ppx/ppx", "--exclusions", "bisect.exclude"]]
```

where got errors:

```sh
> relude@0.61.0 build /Users/jihchi/Repos/relude
> bsb -make-world

[11/11] Building src/runtime/bucklescript/runtime-Bisect.cmj
[81/81] Building bastet/src/Test-BsBastet.cmj
[5/5] Building src/jest.cmj
[1/456] Building __tests__/Relude_Eq_test.reast
FAILED: __tests__/Relude_Eq_test.reast
/Users/jihchi/Repos/relude/node_modules/bs-platform/darwin/bsc.exe  -w +A-4-40-42 -warn-error +A -color always -ppx '/Users/jihchi/Repos/relude/node_modules/bisect_ppx/ppx --exclusions bisect.exclude' -bs-no-version-header -bs-super-errors -o __tests__/Relude_Eq_test.reast -bs-syntax-only -bs-binary-ast /Users/jihchi/Repos/relude/__tests__/Relude_Eq_test.re
File "_none_", line 1:
Error: I/O error: bisect.exclude: No such file or directory

  We've found a bug for you!
  /Users/jihchi/Repos/relude/__tests__/Relude_Eq_test.re

  Error while running external preprocessor
Command line: /Users/jihchi/Repos/relude/node_modules/bisect_ppx/ppx --exclusions bisect.exclude '/var/folders/wm/lp071lgd09nb3_mv_9pthcs00000gn/T/camlppx7aa76b' '/var/folders/wm/lp071lgd09nb3_mv_9pthcs00000gn/T/camlppx94f0b2'

( ...omitted repeated error with different re file... ) 
```

And then, I changed to absolute path:

```js
"ppx-flags": [
  ["bisect_ppx/ppx", "--exclusions", "/Users/jihchi/Repos/relude/bisect.exclude"]
]
```

where got another errors:

```sh
> relude@0.61.0 build /Users/jihchi/Repos/relude
> bsb -make-world

[11/11] Building src/runtime/bucklescript/runtime-Bisect.cmj
[81/81] Building bastet/src/Test-BsBastet.cmj
[5/5] Building src/jest.cmj
[1/456] Building __tests__/Relude_Eq_test.reast
FAILED: __tests__/Relude_Eq_test.reast
/Users/jihchi/Repos/relude/node_modules/bs-platform/darwin/bsc.exe  -w +A-4-40-42 -warn-error +A -color always -ppx '/Users/jihchi/Repos/relude/node_modules/bisect_ppx/ppx --exclusions /Users/jihchi/Repos/relude/bisect.exclude' -bs-no-version-header -bs-super-errors -o __tests__/Relude_Eq_test.reast -bs-syntax-only -bs-binary-ast /Users/jihchi/Repos/relude/__tests__/Relude_Eq_test.re
open Jest
open Expect
open Relude.Globals
[@@@coverage exclude_file]
;;afterAll Bisect.Runtime.write_coverage_data
( ... omitted re content ...)

File "/var/folders/wm/lp071lgd09nb3_mv_9pthcs00000gn/T/camlppx931ed4", line 1:
Error: I can't decide whether /var/folders/wm/lp071lgd09nb3_mv_9pthcs00000gn/T/camlppx931ed4 is an implement
ation or interface file

  We've found a bug for you!
  /Users/jihchi/Repos/relude/__tests__/Relude_Eq_test.re

  Error while running external preprocessor
Command line: /Users/jihchi/Repos/relude/node_modules/bisect_ppx/ppx --exclusions /Users/jihchi/Repos/relude/bisect.exclude '/var/folders/wm/lp071lgd09nb3_mv_9pthcs00000gn/T/camlppx08d86e' '/var/folders/wm/lp071lgd09nb3_mv_9pthcs00000gn/T/camlppx931ed4'

( ...omitted repeated error with different re file... ) 
```

Reply:
These are probably two issues:

- BuckleScript is running Bisect from some unexpected directory (Bisect does support relative paths).
- There is something wrong with assumptions Bisect makes about the order of arguments coming from BuckleScript.

Could you push your latest code, that triggers these errors, into this branch, and I'll fix as much of this as I can in Bisect?

Reply:
I've pushed latest code into this branch. please let me know if anything that I can help.

Reply:
Ok https://github.com/aantron/bisect_ppx/issues/310 and https://github.com/aantron/bisect_ppx/issues/311 fixed it at least in my testing. I'm waiting for CI to build binaries, and then I'll push release 2.3.2 to npm. I looked at the report, and it looks like test cases were excluded successfully. bs-bastet is still included, but that's waiting on BuckleScript.

Reply:
2.3.2 is in npm with the fixes.

Reply:
> I looked at the report, and it looks like test cases were excluded successfully.

I guess that's not what I checked, since the exclude attributes were still inside each test file :/

Reply:
> > I looked at the report, and it looks like test cases were excluded successfully.
>
> I guess that's not what I checked, since the exclude attributes were still inside each test file :/

`--exclusions` works with version `>=2.3.2`. thanks for speedy fixing on bisect ppx!

Reply:
BTW, I was wrong about `bisect.exclude`, it should be `.re` extension:

https://github.com/reazen/relude/blob/777d04d279cf1ec4a5bc54c432b83a486e74d569/bisect.exclude#L1-L2
---
Latest code coverage report: https://coveralls.io/builds/30182377